### ISSUE TITLE
The very old @types/react in this package causes dependency issues

### ISSUE THREAD

**Issue by kaiyoma** [Monar/react-immutable-pure-component#30]
This package is being installed in my project as an indirect dependency and its very old `@types/react` version is causing type errors. Is it possible to use `^` notation instead for the package version? Or at least upgrade to a more recent version and publish a new version?

**Comment by Monar** [comment]
Sure thing, I take a look at it and update, clean package.json

**Comment by Monar** [comment]
@kaiyoma Strange issue you have when @types/react is declared only in devDependencies? So this should not affect you neither when this package is direct nor indirect dependency in your project. Can you copy paste the error, and show me your yarn.lock or package-lock.json fragment regarding dependencies against this package so maybe I can help with your error.

**Comment by kaiyoma** [comment]
I did a "clean install" of all my packages and the errors have gone away, so I think everything's okay now. That said, I will point out that all my other packages are using the latest version of React types (16.9.56) and only react-immutable-pure-component is using an old version:

```
$ yarn why @types/react
yarn why v1.22.5
[1/4] Why do we have the module "@types/react"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@types/react@16.9.56"
info Has been hoisted to "@types/react"
info Reasons this module exists
   - Specified in "devDependencies"
   - Hoisted from "@types#enzyme#@types#react"
   - Hoisted from "@types#react-color#@types#react"
   - Hoisted from "@types#react-copy-to-clipboard#@types#react"
   - Hoisted from "@types#react-dom#@types#react"
   - Hoisted from "@types#react-grid-layout#@types#react"
   - Hoisted from "@types#react-list#@types#react"
   - Hoisted from "@types#react-redux#@types#react"
   - Hoisted from "@types#react-virtualized#@types#react"
   - Hoisted from "@types#recompose#@types#react"
   - Hoisted from "@vx#axis#@types#react"
   - Hoisted from "@vx#gradient#@types#react"
   - Hoisted from "@vx#pattern#@types#react"
   - Hoisted from "@vx#responsive#@types#react"
   - Hoisted from "@vx#stats#@types#react"
   - Hoisted from "@vx#text#@types#react"
   - Hoisted from "@types#enzyme-adapter-react-16#@types#enzyme#@types#react"
   - Hoisted from "@types#react-redux#@types#hoist-non-react-statics#@types#react"
   - Hoisted from "@storybook#api#@types#reach__router#@types#react"
   - Hoisted from "antd#@types#react-slick#@types#react"
   - Hoisted from "@storybook#components#@types#react-syntax-highlighter#@types#react"
   - Hoisted from "@types#react-color#@types#reactcss#@types#react"
   - Hoisted from "@vx#axis#@vx#group#@types#react"
   - Hoisted from "@vx#stats#@vx#group#@types#react"
   - Hoisted from "@vx#axis#@vx#shape#@types#react"
   - Hoisted from "@storybook#core#@storybook#ui#@types#markdown-to-jsx#@types#react"
   - Hoisted from "@testing-library#react-hooks#@types#testing-library__react-hooks#@types#react-test-renderer#@types#re
act"
   - Hoisted from "@sambego#storybook-state#@storybook#components#@types#react-textarea-autosize#@types#react"
info Disk size without dependencies: "172KB"
info Disk size with unique dependencies: "1.28MB"
info Disk size with transitive dependencies: "1.28MB"
info Number of shared dependencies: 2
=> Found "react-immutable-pure-component#@types/react@16.4.6"
info This module exists because "swagger-ui-react#react-immutable-pure-component" depends on it.
info Disk size without dependencies: "128KB"
info Disk size with unique dependencies: "1.22MB"
info Disk size with transitive dependencies: "1.22MB"
info Number of shared dependencies: 1
Done in 2.13s.
```

Maybe it would be good to change your dependency to use `^` notation? It would be good package hygiene to keep up with the latest types.

**Comment by Monar** [comment]
So I got interested in your issue and this is what I found:
If you install latest swagger-ui it depends on  `^1.1.1` of this package: https://github.com/swagger-api/swagger-ui/blob/master/package.json#L83 so with all the `^` logic the varsion `1.2.3` is installed that I was having `@types/react: * ` (as you seen more loose it coudn't have been) in `optionalDependencies` that in `swagger-ui` `package-lock.json` ([here](https://raw.githubusercontent.com/swagger-api/swagger-ui/master/package-lock.json)) was frozen to: 
```
    "react-immutable-pure-component": {
      "version": "1.2.3",
      "requires": {
        "@types/react": "16.4.6"
      }
    },
```
So when you install you drag `package-lock.json` version from `swagger-ui`. 

Something I can do is to release version v1 (in v2 I already dont' have this) without this `optionalDependency` but this will help only when `swagger-ui` in `package-lock.json` will be rebuild. 

And regarding `^` as I've mentioned I don't have it in `dependences` and my `peerDepenedecies` are `*` in dependencies in package and I personally think that having `^` is generally bad idea as you hide specific version in lock files so just gain hidden information and just leave you with less control and rare cases of mismatching install as regeneration of lockfiles is a bit more complex, but this is just short remark and topic for separate discussion. 

To sum up, I'll prepare version for version `v1`, without this `optionalDependency` but when `package-lock.json` in `swagger-ui` will be regenerated I don't know but at some day this should be solved.

**Comment by Monar** [comment]
https://www.npmjs.com/package/react-immutable-pure-component/v/1.2.4 published

**Comment by kaiyoma** [comment]
Very nice, this new version is working out very well in my project. Thanks so much for the attention over the weekend!
